core/tests/unit/ldml/test_context_normalization.cpp

```diff
@@ -0,0 +1,162 @@
+  km_core_cp* new_cached_context = new km_core_cp[buf_size];
+  try_status(km_core_context_items_to_utf16(citems, new_cached_context, &buf_size));
+
+  km_core_context_items_dispose(citems);
+
+  bool result = std::u16string(cached_context) == new_cached_context;
+  delete[] new_cached_context;
+  return result;
+}
+
+void test_context_normalization_already_nfd() {
+  km_core_cp const *app_context_nfd = u"A\u0300";
+  setup("k_001_tiny.kmx");
+  assert(km_core_state_context_set_if_needed(test_state, app_context_nfd) == KM_CORE_CONTEXT_STATUS_UPDATED);
+  assert(is_identical_context(app_context_nfd, KM_CORE_DEBUG_CONTEXT_APP));
+  assert(is_identical_context(app_context_nfd, KM_CORE_DEBUG_CONTEXT_CACHED));
+  teardown();
+}
+
+void test_context_normalization_basic() {
+  km_core_cp const *application_context = u"This is a test À";
+  km_core_cp const *cached_context =      u"This is a test A\u0300";
+  setup("k_001_tiny.kmx");
+  assert(km_core_state_context_set_if_needed(test_state, application_context) == KM_CORE_CONTEXT_STATUS_UPDATED);
+  assert(is_identical_context(application_context, KM_CORE_DEBUG_CONTEXT_APP));
+  assert(is_identical_context(cached_context, KM_CORE_DEBUG_CONTEXT_CACHED));
+  teardown();
+}
+
+void test_context_normalization_hefty() {
+                                          // Latin     Latin            "ṩ"                   "Å"                 Tirhuta U+114bc -> U+114B9 U+114B0
+  km_core_cp const *application_context = u"À"       u"é̖"             u"\u1e69"              u"\u212b"           u"\U000114BC";
+  km_core_cp const *cached_context =      u"A\u0300" u"e\u0316\u0301" u"\u0073\u0323\u0307"  u"\u0041\u030a"     u"\U000114B9\U000114B0";
+  setup("k_001_tiny.kmx");
+  assert(km_core_state_context_set_if_needed(test_state, application_context) == KM_CORE_CONTEXT_STATUS_UPDATED);
+  assert(is_identical_context(application_context, KM_CORE_DEBUG_CONTEXT_APP));
+  assert(is_identical_context(cached_context, KM_CORE_DEBUG_CONTEXT_CACHED));
+  teardown();
+}
+
+void test_context_normalization_invalid_unicode() {
+                                          // unpaired surrogate     illegal
+  km_core_cp const application_context[] = { 0xDC01, 0x0020, 0x0020, 0xFFFF, 0x0000 };
+  km_core_cp const cached_context[] =      { 0xDC01, 0x0020, 0x0020, 0xFFFF, 0x0000 };
+  setup("k_001_tiny.kmx");
+  assert(km_core_state_context_set_if_needed(test_state, application_context) == KM_CORE_CONTEXT_STATUS_UPDATED);
+  assert(is_identical_context(application_context, KM_CORE_DEBUG_CONTEXT_APP));
+  assert(is_identical_context(cached_context, KM_CORE_DEBUG_CONTEXT_CACHED));
+  teardown();
+}
+
+void test_context_normalization() {
+  test_context_normalization_already_nfd();
+  test_context_normalization_basic();
+  test_context_normalization_hefty();
+  // TODO: we need to strip illegal chars: test_context_normalization_invalid_unicode(); // -- unpaired surrogate, illegals
```

Comment:
no function for that yet but could use the new-ish macros.

Reply:
See #10392